### ISSUE TITLE
Export to SVG/PNG/JPEG doesn't work

### ISSUE THREAD

**Issue by hobyst** [penpot/penpot#1050]
**Describe the bug**
None of the export options work and a push notification message appears saying "An unexpected error occurred".

**To Reproduce**
1.  Open any project
2. Select any kind of element and try to export it

**Expected behavior**
After a short amount of time, the browser should prompt the user to save/download the exported file.

**Screenshots**

**Desktop (please complete the following information):**

- Windows 10 Home 21H1
- Firefox 89.0.1
- Penpot Frontend v1.6.5
- Penpot Backend v1.6.5
- Penpot Exporter v 1.6.5

**Environment:**
Self-hosted instance using Docker Engine for WSL 2 running on Windows 10 Home 21H1 (x64).

-   Docker version 20.10.6, build 370c289

**Frontend Stack Trace**
```
172.18.0.2 - - [19/Jun/2021:20:18:40 +0000] "GET / HTTP/1.1" 200 114611 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:40 +0000] "GET /css/main-default.css?ts=1623932894408 HTTP/1.1" 200 37656 "http://penpot-frontend/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:40 +0000] "GET /js/config.js?ts=1623932894634 HTTP/1.1" 200 248 "http://penpot-frontend/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:40 +0000] "GET /js/polyfills.js?ts=1623932894634 HTTP/1.1" 200 1004 "http://penpot-frontend/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:40 +0000] "GET /js/shared.beaea732.js HTTP/1.1" 200 400474 "http://penpot-frontend/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:40 +0000] "GET /js/main.beaea732.js HTTP/1.1" 200 880797 "http://penpot-frontend/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:41 +0000] "GET /js/worker.beaea732.js HTTP/1.1" 200 12606 "http://penpot-frontend/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:41 +0000] "GET /fonts/WorkSans-Regular.woff HTTP/1.1" 200 98800 "http://penpot-frontend/css/main-default.css?ts=1623932894408" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:41 +0000] "GET /fonts/WorkSans-Light.woff HTTP/1.1" 200 103552 "http://penpot-frontend/css/main-default.css?ts=1623932894408" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.2 - - [19/Jun/2021:20:18:41 +0000] "GET /fonts/WorkSans-ExtraLight.woff HTTP/1.1" 200 103992 "http://penpot-frontend/css/main-default.css?ts=1623932894408" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"

172.18.0.1 - - [19/Jun/2021:20:19:10 +0000] "POST /export HTTP/1.1" 500 0 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0"
```

**Backend Stack Trace**

Nothing remarkable.

**Exporter stack trace**
Right after a PNG or a JPEG export request the following line is logged:
```
[app.renderer.bitmap] {:uri "http://penpot-frontend/#/render-object/<WORKSPACE_ID>/<PROJECT_ID>/<PAGE_ID>", :line 42}
```
Right after a SVG export request  the following line is logged:
```
[app.renderer.svg] {:uri "http://penpot-frontend/#/render-object/<WORKSPACE_ID>/<PROJECT_ID>/<PAGE_ID>", :line 281}
```
**Note:** `<WORKSPACE_ID>`, `<PROJECT_ID>` and `<PAGE_ID>` represent redacted content and in the real logs these spaces are filled with identificators.

**Comment by Shriukan33** [comment]
Can't export from browser either : 
Firefox 89.0 (64-bit)
Ubuntu 20.04

Push notification appears after few seconds "An unexpected error occurred".

**Comment by elhombretecla** [comment]
Hi all!!!

We will check this bug asap. It seems to be related to the local environment configuration but we need to do more research.

Sorry for the inconvenince

**Comment by niwinz** [comment]
@Shriukan33 I'm not clearly understand that you mean with "can't export from browser either", because there are no other way to do the export.

@hobyst  Looking at the http server log, the export operation returns 500, this should print some stack trace on exporter side log. In any case WSL2 is not a supported platform, so there are some changes that this issue is caused by some incompatibilities that we are not aware right now. Unfortunately without a detailed stack-trace or some other additional info that can help us understand that is causing this error, it is very complicated to guess the fix for it.

**Comment by b0oml** [comment]
Hi, I'm also getting this issue.

My instance is self-hosted using Docker and docker-compose on CentOS 8.
I'm using this [docker-compose.yml](https://raw.githubusercontent.com/penpot/penpot/main/docker/images/docker-compose.yaml) and this [config.env](https://raw.githubusercontent.com/penpot/penpot/main/docker/images/config.env) (only modifications are PENPOT_PUBLIC_URI, SMTP and LDAP).

Here is the exporter stack trace when trying to export something as PNG or JPG:

```
[app.http.impl] {:msg "Unexpected error", :error #object[Error Error: Timeout hit: 30000], :line 49}
Error: Timeout hit: 30000
    at Object.<anonymous> (/opt/app/node_modules/puppeteer-cluster/dist/util.js:69:23)
    at Generator.next (<anonymous>)
    at fulfilled (/opt/app/node_modules/puppeteer-cluster/dist/util.js:5:58)
```

The stack strace is a bit different when exporting as SVG:

```
[app.http.impl] {:msg "Unexpected error", :error #object[TimeoutError TimeoutError: waiting for selector `#screenshot` failed: timeout 15000ms exceeded], :line 49}
TimeoutError: waiting for selector `#screenshot` failed: timeout 15000ms exceeded
    at new WaitTask (/opt/app/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:509:34)
    at DOMWorld.waitForSelectorInPage (/opt/app/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:420:26)
    at Object.internalHandler.waitFor (/opt/app/node_modules/puppeteer/lib/cjs/puppeteer/common/QueryHandler.js:31:77)
    at DOMWorld.waitForSelector (/opt/app/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:313:29)
    at Frame.waitForSelector (/opt/app/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:841:51)
    at Page.waitForSelector (/opt/app/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1333:33)
    at Function.app.browser.wait_for.cljs$core$IFn$_invoke$arity$3 (/opt/app/app.js:48963:18)
    at Function.app.browser.wait_for.cljs$core$IFn$_invoke$arity$2 (/opt/app/app.js:48957:31)
    at /opt/app/app.js:55680:63
    at /opt/app/app.js:31726:108
```

**Comment by Shriukan33** [comment]
> @Shriukan33 I'm not clearly understand that you mean with "can't export from browser either", because there are no other way to do the export.

I was just precising that I'm not using a docker container, just plain old Firefox and Ubuntu.

**Comment by niwinz** [comment]
@b0oml hmm strange, but is something that is already happened to us, the exporter process uses headless browser for rendering, and sometimes it just breaks, and strarts to timeout. My recommendation is just restart the process.
In any case, tomorrow I try to test/check the latest version published on docker if something is wrong.

**Comment by niwinz** [comment]
> > @Shriukan33 I'm not clearly understand that you mean with "can't export from browser either", because there are no other way to do the export.
> 
> I was just precising that I'm not using a docker container, just plain old Firefox and Ubuntu.

Yeah, this is the way you do it using docker container or not. So I'm still not clearly understand. Can you specify if you are use our SAAS (at https://design.penpot.app) or you on premise instance. In case of SAAS, send us at `info@penpot.app` describing the problem, a link to the file (a url when you are on the workspace on the file you try to export) and we try to investigate.

**Comment by Shriukan33** [comment]
> > > @Shriukan33 I'm not clearly understand that you mean with "can't export from browser either", because there are no other way to do the export.
> > 
> > 
> > I was just precising that I'm not using a docker container, just plain old Firefox and Ubuntu.
> 
> Yeah, this is the way you do it using docker container or not. So I'm still not clearly understand. Can you specify if you are use our SAAS (at https://design.penpot.app) or you on premise instance. In case of SAAS, send us at `info@penpot.app` describing the problem, a link to the file (a url when you are on the workspace on the file you try to export) and we try to investigate.

I was using SAAS at the time. 
Didn't try again since then, but it seems to work today. Maybe something went wrong at the time, seems fixed for now. 

Sorry for inconvenience.

**Comment by DanielOaks** [comment]
For what it's worth I'm getting the same stack trace from my exporter container that @b0oml is, and get the same even if I've just restarted that container. My setup's running on Arch+Docker. I'll see if I can get any more info, and if I do will post it in here. (I only started using penpot yesterday so it wasn't working before and only broken now, etc)

~~**edit:** I'm not familiar with Clojure at all but maybe it's worth looking at how `public-uri` is defined [here in the frontend](https://github.com/penpot/penpot/blob/432a8f233838099eff0b6146e5558620625f4211/frontend/src/app/config.cljs#L92) vs [here in the exporter](https://github.com/penpot/penpot/blob/e3171d9ee592370dc61bd084706d966a398ecee6/exporter/src/app/config.cljs#L24)? The frontend one refers to `penpotPublicURI` but the exporter one doesn't, maybe that could have something to do with things?~~

**edit:** From the looks of it, mine seems to be hanging on this line, and it doesn't get to the `p/let case` line below it before the 30s timeout is up: https://github.com/penpot/penpot/blob/develop/exporter/src/app/renderer/bitmap.cljs#L52
I'll keep digging and see if I can find any more useful details.

**Comment by DanielOaks** [comment]
puppeteer-cluster also [uses a fairly old version of puppeteer](https://github.com/thomasdondorf/puppeteer-cluster/pull/399), I wonder if whatever's causing could be a puppeteer issue that's been solved in the meantime, or something to do with incompatibilities? I've tried navigating to the url that it uses to render and it seems like that's being sent fine with a `#screenshot` element, hmm.

**Comment by niwinz** [comment]
I have plans to completelly replace puppeter-cluster, but right now is not the issue, in fact we use the latest puppeter version (we have pinned it.

I have published image with `1.7.1-alpha` and `latest` with new images that fixes a configuration issue related to exporter timeouts. It should fix the issue on the docker instances.

**Comment by DanielOaks** [comment]
Ah perfect, thank you very much for the fix! Works perfectly on my end now \o/
